### DISCUSSION TITLE
Version 2.2 Beta 1

### DISCUSSION THREAD

**Discussion by vkbo** [vkbo/novelWriter#1596] [General]
### Release Notes

This is a beta release of the next release version, and is intended for testing purposes. Please be careful when using this version on live writing projects, and make sure you take frequent backups.

Please check the changelog for an overview of changes. The full release notes will be added to the final release.

### Detailed Changelog

**Features**

* novelWriter has a new logo and icon. PR #1593.
* The Document Editor is now a true plain text editor. This has a number of benefits and a couple of drawbacks. The most important benefits is that the editor responds a lot faster, and can hold much larger text documents. The big document limit has therefore been removed. It mostly affected automatic spell checking. The syntax highlighter and spell checker are also more efficient, which allows for needed improvements to these. The drawbacks are mainly that the editor now scrolls one line at a time, instead of scrolling pixel by pixel like before. PRs #1521 and #1525.
* Tags and References are now case insensitive. Their display name on the user interface remains the same as the value set for the `@tag` entry. Issue #1313. PRs #1522 and #1578.
* Keywords for Tags and References, and the References themselves, now have an auto-complete menu that pops up in the editor on lines starting with the `@` character. It will first suggest what keyword you want to use, and when it has been added, use that keyword to look up suggestions for references to add. The suggestions improve as you type by looking for the characters you've already typed in the tags you've previously defined. Issue #823. PR #1581.
* You can now right-click an undefined tag, and a context menu option to create a Project Note for that tag will appear in the menu. On selection, it will create a note in the first root folder of the correct kind, and set the title and tag to match the undefined reference, making it instantly defined. Issues #1580 and #823. PR #1582.
* Shortcodes have been added to the Document Editor. Shorcodes are HTML-like syntax, but uses square brackets instead of angular brackets. So `[b]text[/b]` will make the word "text" appear as bold. Shortcodes currently support bold, italic, striketrough, underline, superscript and subscript text. The first three are complimentary to the Markdown-like syntax that. The benefit of the shortcode emphasis syntax, however, is that it does not care about word boundaries, and can therefore be used any place in the text. Including in the middle of words. Issues #1337 and #1444. PRs #1540 and #1583.
* A show/hide toolbar has been added to the editor where toolbuttons for formatting options are available. The toolbar is hidden by default, but can be activated from a three dots icon in the top left corner of the editor. Issue #1585. PR #1584.
* Build Definitions in the Manuscript Build tool can now be re-ordered, and the order is preserved when the tool is closed and re-opened. Issue #1542. PR #1591.

**Usability**

* The Settings menu in the sidebar now always pops out to the right and upwards from the bottom of the icon. The previous behaviour was not guaranteed to stay in the visible area of the screen. PR #1520.
* The right click action on a misspelled word now uses the actual spell checker data for lookup. Previously, the spell checker would underline a word that was misspelled, but the right click action actually had no way of reading where the error line was, so it had to guess again what word the user was clicking. Since these two parts of the code used different logic, they sometimes produced different results. The spell checker now saves the location of each spell check error, and the right click action retrieves this data when generating suggestions, which should eliminate the problem of picking the correct word boundaries. Issue #1532. PR #1525.
* The language of a project is not set in the New Project Wizard and in Project Settings. It is no longer defined in the Build Settings panel. Issue #1588. PR #1589.
* The way switching focus and view in the main GUI has changed. Pressing `Ctrl+T` will now switch focus to the Project or Novel Tree if focus is elsewhere, or if either have focus already, it will switch view to the other tree. Pressing `Ctrl+E` will switch focus and view to the Document Editor. Pressing `Ctrl+Shift+T` will do the same for the Outline View. The old Alt-based shortcuts have been removed. Issues #1310 and #1291. PR #1590.

**User Interface**

* The labels under the sidebar buttons have been removed. The tool tips have the necessary information. PR #1520.

**Other Improvements**

* Also the Tags and References keywords are now translated into the project language when these are included in Manuscript builds. As long as the phrases have been translated. PR #1586.

<hr /><em>This discussion was created from the release <a href='https://github.com/vkbo/novelWriter/releases/tag/v2.2b1'>Version 2.2 Beta 1</a>.</em>

**Comment by HeyMyian** [comment]
Clicking on the formatting buttons in the new toolbar currently does nothing when
1) no word is selected or
2) the cursor is not inside of or at the end of a word.

The expected result however is to insert the formatting symbols with the cursor placed inside them, ready to type.

**Reply by vkbo** [reply]
This is the exact same functionality that the Format menu has (it's the same code). But this behaviour makes sense when you have a toolbar, I agree.

Could you make a feature request?

**Comment by HeyMyian** [comment]
> The labels under the sidebar buttons have been removed.

Not sure if this was already the case in previous versions, but now that the sidebar is as wide as the icons need, it's noticeable: There is a 2px right-border or padding on the sidebar that has the same color as the sidebar background. Thus icons appear to not be horizontally centered on the sidebar. Here is a zoomed screenshot that shows the issue:

![2023-11-12___18_44_49](https://github.com/vkbo/novelWriter/assets/60368695/dd3484c9-4427-4736-89f6-0fd8d72ed827)

I checked the other two programs that I have that I know use a similar sidebar. VSCode does not have a border (no hover effect), and some Qt app uses a slightly darker, 1px width border.

![2023-11-12___18_49_15](https://github.com/vkbo/novelWriter/assets/60368695/4cf41ff6-4505-4c6e-8c79-d9c5ca687dfa)
![2023-11-12___18_42_49](https://github.com/vkbo/novelWriter/assets/60368695/876ed855-91bc-44b7-b476-28b1988f492e)

**Reply by vkbo** [reply]
On my desktop I have a margin to the left of the sidebar, so I added the 2 pixels to actually make it centre, but I suppose we're on the window manager real estate, which the app doesn't control. Either way, it will look wrong on some desktops and right on some. Not sure what to do about it.

**Reply by HeyMyian** [reply]
Where does your left margin actually come from? On Linux (AppImage) I simply have my normal window border on the left side (1px blue). I use "Breeze" window decorations with window border size "no borders".
![Screenshot_2023-11-12_21-11-59](https://github.com/vkbo/novelWriter/assets/60368695/fb2e3208-bf15-4c69-8d75-5737f423b189)

**Reply by vkbo** [reply]
It comes from the window manager or desktop. It lies outside the area novelWriter draws on. If I take the width and height of the app window as novelWriter sees it, it is smaller than the size of the app in the desktop environment (I've measured on screenshots too). Qt has very little control of what happens there as it's sort of out of reach. I've run into this in a few cases. From your screenshot, the light blue border and the title bar are all part of this desktop real estate.

**Reply by HeyMyian** [reply]
I installed a different "Window Decoration" to test. This one is called "Breezemite" and uses "normal" window border size. Is it something like this for you? If so, I don't think novelWriter should account for that.

![Screenshot_2023-11-12_21-29-38](https://github.com/vkbo/novelWriter/assets/60368695/de5e47cc-f294-4442-ab9d-bd6a951423a5)

![Screenshot_2023-11-12_21-36-58](https://github.com/vkbo/novelWriter/assets/60368695/64a1e21c-bd0c-4f11-802b-719d92762434)

**Reply by vkbo** [reply]
That looks similar, yes. Anyway, I've removed the margin in #1601, together with a usability fix for the tag/reference completer,

**Comment by HeyMyian** [comment]
> The Document Editor is now a true plain text editor. [...] The drawbacks are mainly that the editor now scrolls one line at a time, instead of scrolling pixel by pixel like before.

The editor now has a vertical scrollbar as soon as it contains more than one line of text. Not sure if intended behavior, so I just wanted to notify you of it for now. I suspect in previous versions, the appearance of the scrollbar was tied to the setting "Scroll past end of the document    x lines", which does not exist anymore.

**Reply by vkbo** [reply]
Yes, this is also caused by the centerOnScroll feature mentioned in my other reply. I have no control over the way this setting works, so as I mention in the other thread, all I can do is add a way for the user to choose whether this feature is enabled or not.

You can also disable the vertical scroll bar altogether, which is what I do. I use the mouse wheel for scrolling on my desktop PC anyway.

**Comment by HeyMyian** [comment]
Another observation about this:

> The Document Editor is now a true plain text editor.

When the cursor is in the bottom-most visible line, hitting the down arrow expectedly moves the cursor to the next line. But unexpectedly, the entire viewpoint jumps by half a page into the vertical center. Not sure how to describe. In previous versions, the editor would move downward line by line. I think this new behavior is unusual and disorienting.

**Reply by vkbo** [reply]
This is the centerOnScroll feature. It is a Qt feature and not something I added. I can turn it on or off, but if it is off, you can also not scroll past the end of the document. The old implementation that let you do that does not work in the plain text editor. The best I can do is add back in the option in Preferences and use it to enable/disable this feature.

Edit: Here is its description https://doc.qt.io/qt-5/qplaintextedit.html#centerOnScroll-prop

**Reply by HeyMyian** [reply]
Thanks. Oof, that's a harsh tradeoff, not being able to scroll past the end of the document. Is there not some kind of magic you could do with bottom padding/margin to "fake" scrolling past the document end?

> The best I can do is add back in the option in Preferences and use it to enable/disable this feature.

I would appreciate the toggle.

**Reply by vkbo** [reply]
No, unfortunately not. that's how I achieved it in the old implementation. I just added a bottom margin on the root frame of the document.

The plain text editor is optimised for plain text, so it works on the assumption of fixed lines. It is missing a lot of the layout features of the rich text editor, but it is a lot more efficient since it doesn't have to account for a lot of things you may want to render, like lists, images and tables and such. It doesn't matter that novelWriter didn't use any of these features. They are there in the layout engine anyway.

The scrolling effects is the main tradeoff with the plain text editor.

**Reply by HeyMyian** [reply]
What a bummer from the source. Okay.

I assume the "typewriter scroll" (I never used it) also wouldn't work anymore if centerOnScroll is false?

**Reply by HeyMyian** [reply]
I love sleek, lean, efficient things btw so I understand why you changed it to plain text :smile:

**Reply by vkbo** [reply]
Typewriter scroll is completely my code and would be unaffected. It detects when the cursor changes line due to a key press (except movement keys) and then scrolls a line up or down to keep the cursor at the same vertical position. This works like before, but due to the plain text editor, it skips a line rather than scrolling smoothly in 200ms like the old implementation.

I added it because it is a lot more comfortable when writing, and it avoids the need to scroll while doing so.

**Reply by vkbo** [reply]
> I love sleek, lean, efficient things btw so I understand why you changed it to plain text 😄

The text editor used to have a lot of checks and hacks to handle various tasks on large documents. Particularly annoying was the refresh on large documents, like when you manually run the spell checker or change any of the preferences that affects the editor. In those cases, the only way to refresh efficiently was to reload the text. This clears the undo buffer.

The syntax highlighter and spell checking was also rather slow. This is not noticeable on a small document of a couple thousand words. But I have a stress test document of 800k+ words. It took about 10 minutes to refresh. Now it takes 3 seconds. It took a lot less time to load though, and I never managed to figure out why the difference. I even discussed it with some of Qt's developers once. It definitely has to do with the layout engine somehow.

**Reply by vkbo** [reply]
Another complicated and fragile hack was the restore cursor position on long documents. Due to the very slow rich text layout engine, the cursor move to restore the position in the document from last time you saved would often trigger before the area to jump to even existed. Hence it wouldn't move at all and just silently fail. The hack was to run a timer loop to poll how far the engine had gotten every few milliseconds until the estimated cursor position was rendered in the buffer, and then make the jump.

That is no longer needed either, and that whole messy code is gone.

You may have noticed the esoteric "Big document limit" setting in old Preferences. It no longer exists in 2.2 because it isn't needed. A bunch of hardcoded limits and thresholds have also been removed. They were always hard to guess where to put, because there was likely a difference to what was "too much" on different OSes and hardware. I tuned it to my PC running Linux on hardware from 2017.

**Comment by vkbo** [comment]
For the 2.2 Release Candidate, I also plan to have the new Tags and References panel ready. Here's a quick demonstration of what I've got so far.

https://github.com/vkbo/novelWriter/assets/1619840/0626f433-02da-42c8-8447-f444d49e499c
